atc/atccmd/command.go

```diff
@@ -218,6 +229,9 @@ type RunCommand struct {
 	EnableRedactSecrets bool `long:"enable-redact-secrets" description:"Enable redacting secrets in build logs."`
 
 	ConfigRBAC string `long:"config-rbac" description:"Customize RBAC role-action mapping."`
+
+	SystemClaimKey    string   `long:"system-claim-key" default:"sub" description:"The token claim key to use when matching system-claim-values"`
+	SystemClaimValues []string `long:"system-claim-value" default:"ChBjb25jb3Vyc2Utd29ya2VyEgZjbGllbnQ" description:"Configure which token requests should be considered 'system' requests."`
```

Comment:
What this configuration is saying is any token with a `sub` equal to `ChBjb25jb3Vyc2Utd29ya2VyEgZjbGllbnQ` should be considered a system token.

You should never have to change this value unless you change this:
https://github.com/concourse/concourse/blob/a8bcb4a8bb2f6f195f018e17d53b32eaf2c32919/tsa/tsacmd/command.go#L47

A responsible concourse operator only really needs to change the `CONCOURSE_TSA_CLIENT_SECRET` but we might have clients that want to change `CONCOURSE_TSA_CLIENT_ID` just for some added peace of mind.

So where does this value come from?

A token issued by dex to the tsa using the client credentials grant type looks like this:
```
{
   "iss":"http://localhost:8080/sky/issuer",
   "sub":"ChBjb25jb3Vyc2Utd29ya2VyEgZjbGllbnQ",
   "aud":"concourse-worker",
   "exp":1585664101,
   "iat":1585577701,
   "at_hash":"HucHOAM3G3tzgI-jhm_a-A"
}
```

The `sub` is the single unique identifier for this client/grant type combination. It ends up being a weird value because dex needs to maintain a unique namespace across all connectors so it hashes the user_id and the connector_id to get a sub thats guaranteed to be unique across all connectors. (All client credentials are granted under a `client` connector)

Instead of using the `sub` we could use the `aud` which will always have a human friendly value, however the `aud` doesn't change based on the grant type used to login. So if someone were to login as a user, then the `sub` would be different but the `aud` would always be `concourse-worker`. This might be fine though since this client is only ever use by the tsa for one specific purpose?

Reply:
Gotcha, thanks for the background on that - I was wondering why the base64-encoded value looked so strange!

How would you feel about automatically deriving this value based on the configured `ClientID`? Basically just doing the `base64(ClientID+"client")` ourselves. The trade-off is that it's tightly coupled to Dex internals, but we at least wouldn't have to write awkward documentation around it. 🤷‍♂

Come to think of it - if Dex ever changes how this value is created, it would probably be less painful to change our tightly coupled code rather than update the documentation and instruct users to generate the value in the new way. 🤔 

wdyt?

Reply:
We definitely could. My one hesitation is that I've been slowly trying to inch our configuration to a place where we could theoretically deploy concourse without dex. Like being able to point to an existing OIDC provider. In that world, adding these configurations that let you describe a 'system' request as a claim key/value pair is really useful.

What are your thoughts on changing the defaults to:
`CONCOURSE_SYSTEM_CLAIM_KEY=aud`
`CONCOURSE_SYSTEM_CLAIM_VALUE=concourse-worker`

Reply:
@pivotal-jwinters That sounds more usable, but would that also mean that if someone e.g. logs in via GitHub with a `concourse-worker` username, they would effectively have a system token?

Reply:
The `aud` claim is always the client_id you use to login. All web logins use a `concourse-web` client so it would never conflict no matter what their user info.

Reply:
Ah OK. If there's no way to spoof it then yeah, that sounds good to me!

Reply:
The one caveat is that all of this is configureable. So in theory someone could configure the web client and the tsa client to have the same client_id. In that case every token in the system would end up being a system token. Though for this to happen you would have to knowingly change these values, and presumably if you change these values, you know what you're doing.